gcamwrapper/main.py

```diff
@@ -8,41 +9,116 @@ class Gcam (gcam_module.gcam):
        instance.
     """
 
-    def get_data(self, query):
+    def run_to_period(self, period = None):
+        """ Run GCAM up to and including some model period.
+            Model periods which have already been run will be kept track
+            of and will not be run again.  HOWEVER, we do not attempt to
+            keep track of if those model periods are "dirty" such as if
+            a user has called `set_data` in such a way that would invalidate
+            that solution.
+
+        :param period: The model period to run or the `get_current_period` + 1
+                       if None
+        :type period: int
+        """
+
+        if period is None:
+            period = self.get_current_period() + 1
+        super(Gcam, self).run_to_period(period)
+
+    def get_data(self, query, query_params = None):
         """Queries for arbitrary data from a running instance of GCAM.
 
         :param query:   GCAM fusion query
         :type query:    str
+        :param query_params: User options to translate placeholder expressions in query should
+                             it have any
+        :type query_params:  dict(string: array(string))
 
         :returns:       DataFrame with the query results.
 
         """
 
+        units = query.units if hasattr(query, "units") else None
+        if query_params is not None:
+            query = apply_query_params(query, query_params, True)
         data_dict = super(Gcam, self).get_data(query)
         data_df = DataFrame(data_dict)
         # The data comming out of gcam is unaggregated so we will need to do that now
         # first figure out what the "value" column is, group by everything else, and summarize
         # TODO: decide on failure mode when no results
         cols = data_df.columns
         value_col = cols[-2] if cols[-1] == "year" else cols[-1]
-        return data_df.groupby(cols.drop(value_col).to_list(), as_index=False).sum()
+        data_df = data_df.groupby(cols.drop(value_col).to_list(), as_index=False).sum()
+        if units is not None:
+            data_df.meta = {'units': units}
+        return data_df
 
-    def set_data(self, data_df, query):
+    def set_data(self, data_df, query, query_params = None):
         """Changes arbitrary data in a running instance of GCAM.
 
         :param data_df:     DataFrame of data to set
         :type data_df:      DataFrame
         :param query:       GCAM fusion query
+        :param query_params: User options to translate placeholder expressions in query should
+                             it have any
+        :type query_params:  dict(string: array(string))
         :type query:        str
 
         """
 
+        if query_params is not None:
+            query = apply_query_params(query, query_params, False)
         data_dict = dict()
         for key, value in data_df.items():
```

Comment:
You can assign a dictionary directly from a data frame like:

```python
import pandas as pd


df = pd.DataFrame({'a': [1, 2, 3], 'b': [5, 6, 7]})

df.set_index('a', inplace=True)

my_dict = df.to_dict().get('b')
```
which would return `{1: 5, 2: 6, 3: 7}` for `my_dict`

You can also set a default dict value on that `my_dict = df.to_dict().get('b')` call like `my_dict = df.to_dict().get('b', None)` where `None` could be whatever you wanted the default dict to be if `b` was not there when looked up.

Reply:
Hmm, I need something a little different in that I need `to_dict()` but the "values" need to come out as a numpy array, like: `{'a': array([1, 2, 3]), 'b': array([5, 6, 7])}`

I couldn't readily figure out how to get `to_dict()` or the like to give me exactly what I needed so I just went with the naive approach.  Was there some straightforward way to get pandas to do what I need that I missed?

Reply:
Leave as-is in this case for now